### ISSUE TITLE
Potentially unexpected noisy gradient

### ISSUE THREAD

**Issue by NTT123** [ikostrikov/pytorch-a2c-ppo-acktr-gail#176]
https://github.com/ikostrikov/pytorch-a2c-ppo-acktr-gail/blob/8a4a500b8bd7a7c10f7b3fbcaebcfdb6d62ef7d4/a2c_ppo_acktr/storage.py#L122-L125

`drop_last` should be `True`. 

There is a potential case in which the last mini-batch is very small (e.g. of size 1). The gradient will be very noisy and it can destroy the progress of previous steps.

**Comment by NTT123** [comment]
As an example of the noisy gradient, I ran the following command

`python3 main.py --env-name "CartPole-v1" --algo ppo --use-gae --lr 2.5e-4 --clip-param 0.1 --value-loss-coef 0.5 --num-processes 8 --num-steps 128 --log-interval 1 --use-linear-lr-decay --entropy-coef 0.01 --use-proper-time-limits --seed 1 --num-mini-batch 5 --num-env-steps 100000`

We have `batch_size = num_processes * num_steps = 8 * 128 = 1024`.

We set `num-mini-batch = 5` . 

The red line is when `drop_last=False` and the blue line is when `drop_last=True`. 

We observed many spikes in the mean discounted reward in the red line due to noisy gradients. 

![batch5](https://i.imgur.com/azk1MLk.png)

We observed no difference when the `num-mini-batch=4`, this is because 1024 is divisible by 4.

![batch5](https://i.imgur.com/0TpAQXH.png)
